client/client.go

```diff
@@ -484,7 +486,69 @@ func loadClientTrust(ctx *openssl.Ctx, conf *Config) (*openssl.Ctx, error) {
 	return ctx, nil
 }
 
-func dialOpenSSL(ctx *openssl.Ctx, conf *Config, _ string, addr string) (net.Conn, error) {
+func upgradeTCPConnectionToSSL(
+	ctx *openssl.Ctx,
+	conn net.Conn,
+	addr string,
+	flags openssl.DialFlags,
+) (*openssl.Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	tlsConn, err := openssl.Client(conn, ctx)
+	if err != nil {
+		return nil, err
+	}
+	if flags&openssl.DisableSNI == 0 {
+		err = tlsConn.SetTlsExtHostName(host)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+	}
+	err = tlsConn.Handshake()
+	if err != nil {
+		tlsConn.Close()
+		return nil, err
+	}
+	if flags&openssl.InsecureSkipHostVerification == 0 {
+		err = tlsConn.VerifyHostname(host)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+	}
+	return tlsConn, nil
+}
+
+func proxyDialOpenSSL(
+	ctx *openssl.Ctx,
+	conf *Config,
+	network, addr string,
+	proxyUrl *url.URL,
+) (net.Conn, error) {
+	flags := openssl.DialFlags(0)
+
+	if conf.NoVerify {
+		flags = openssl.InsecureSkipHostVerification
+	}
+
+	httpProxy := &httpProxyDialer{proxyUrl}
+	proxyConn, err := httpProxy.Dial(network, addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to connect to proxy")
+	}
+
+	conn, err := upgradeTCPConnectionToSSL(ctx, proxyConn, addr, flags)
+	if err != nil {
+		return nil, err
+	}
+
+	return verifyTLSResult(conf, conn)
+}
+
+func dialOpenSSL(ctx *openssl.Ctx, conf *Config, _, addr string) (net.Conn, error) {
```

Comment:
I see you moved out `verifyTLSResult(conf, conn)` for reuse. I think we should do the same for the preamble part (getting the flags) so that we repeat no code between `dialOpenSSL` and `proxyDialOpenSSL`.

Reply:
An alternative is to refactor and merge both functions, just handling when `proxyUrl` is not `nil`. This way the calls from lines 791-794 would look nicer (but there the code will read uglier). Your call.

Reply:
No, I agree with this. Yes would be a bit uglier, but its only a quick branching

Reply:
Yes, if we do the branching we can undo the refactoring of `verifyTLSResult`. I trust your :eye: on this

Reply:
Ah, but nevermind. When refactoring I saw why I did it like this in the first place. `dialOpenSSL` is also used here:
https://github.com/mendersoftware/mender/blob/46a4419ffa64902bd6570b408df9da8734da4309/client/client.go#L563-L570
This would be OK, but doing so will actually solve MEN-6009, which is not the scope of this PR, even thought they are closely related. 
So either do it, then extend the scope of this PR to cover both of the issues, or have something similar to what its like now, then refactor later when solving MEN-6009.

What do you think?

Reply:
I did this, but somehow I liked `verifyTLSResult` also. Im 50/50 split on that